modules/renter/contractor/journal.go

```diff
@@ -0,0 +1,492 @@
+package contractor
+
+// The contractor achieves efficient persistence using a JSON transaction
+// journal. It enables efficient ACID transactions on JSON objects.
+//
+// Each journal represents a single JSON object. The object is serialized as
+// an "initial object" followed by a series of update sets, one per line. Each
+// update specifies a field and a modification. See the journalUpdate type for
+// a full specification.
+//
+// During operation, the object is first loaded by reading the file and
+// applying each update to the initial object. It is subsequently modified by
+// appending update sets to the file, one per line. At any time, a
+// "checkpoint" may be created, which clears the journal and starts over with
+// a new initial object. This allows for compaction of the journal file.
+//
+// In the event of power failure or other serious disruption, the most recent
+// update set may be only partially written. Partially written update sets are
+// simply ignored when reading the journal. Individual updates may also be
+// ignored if they are malformed, though other updates in the set may be
+// applied. See the journalUpdate docstring for an explanation of malformed updates.
+
+// TODO:
+// - handle case sensitivity
+// - handle null vs. empty array
+//    - if index is 0, just do a wholesale replace with [val]
+// - handle inserting new keys into object
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+)
+
+// A journal is a log of updates to a JSON object.
+type journal struct {
+	f        *os.File
+	filename string
+}
+
+// update applies the updates atomically to j. It syncs the underlying file
+// before returning.
+func (j *journal) update(us []journalUpdate) error {
+	buf := make([]byte, 0, 1024) // reasonable guess; avoids GC if we're lucky
+	buf = append(buf, '[')
+	for i, u := range us {
+		if i > 0 {
+			buf = append(buf, ',')
+		}
+		buf = append(buf, `{"p":"`...)
+		buf = append(buf, u.Path...)
+		buf = append(buf, `","v":`...)
+		buf = append(buf, *u.Value...)
+		buf = append(buf, '}')
+	}
+	buf = append(buf, ']', '\n')
+	if _, err := j.f.Write(buf); err != nil {
+		return err
+	}
+	return j.f.Sync()
```

Comment:
This is a future thing, but for max performance:

1. update has to be done under a lock.
2. sync does not have to be done under a lock
3. function should not return until sync is complete

and as a confounding variable, you can't have multiple threads interacting with the same part of the log at the same time, but if you are going to have any good performance they must be able to modify different contractors in parallel. I don't think we are that far along to be worrying about this, but it's on the list of future headaches.

Also this is a great counterargument to the getters/setters thing. The race conditions could get really difficult to manage, especially in a performance sensitive way.